File: OC/OC/HuanView.m
//
//  HuanView.m
//  OC
//
//  Created by STL_QYH on 2023/7/27.
//  Copyright © 2023 STL_. All rights reserved.
//

#import "HuanView.h"

#define kDuration 1.0
#define kDefaultLineWidth 10

@interface HuanView()

@property (nonatomic,strong) CAShapeLayer *backgroundLayer;
@property (nonatomic,strong) CAShapeLayer *progressLayer;
@property (nonatomic,strong) UILabel *progressLabel;
@property (nonatomic,strong) NSTimer *timer;
@property (nonatomic,assign) CGFloat startAngle; // M_PI*2
@property (nonatomic,assign) CGFloat endAngle;

@end


@implementation HuanView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setBackgroundColor:[UIColor clearColor]];
        [self createSubViews];
        //init default variable
        self.backgroundLineWidth = kDefaultLineWidth;
        self.progressLineWidth = kDefaultLineWidth;
        self.percentage = 0;
        self.offset = 0;
//        self.startAngle = -M_PI_2;
//        self.endAngle = 0;
        self.startAngle = -M_PI_2;
        self.endAngle = -M_PI_2-2*M_PI;
    }
    return self;
}

- (void)createSubViews {
    //self.progressLabel.text = @"0%";
    self.progressLabel.textAlignment = NSTextAlignmentCenter;
    //  self.progressLabel.font = FONTBOLD(12);
    [self addSubview:self.progressLabel];
    
    _backgroundLayer = [CAShapeLayer layer];
    _backgroundLayer.frame = self.bounds;
    _backgroundLayer.fillColor = nil;
    _backgroundLayer.strokeColor = [UIColor lightGrayColor].CGColor;
    
    _progressLayer = [CAShapeLayer layer];
    _progressLayer.frame = self.bounds;
    _progressLayer.fillColor = nil;
    _progressLayer.strokeColor = [UIColor redColor].CGColor;
    
    [self.layer addSublayer:_backgroundLayer];
    [self.layer addSublayer:_progressLayer];
}

-(void)setProgressLabelText:(NSString *)progressLabelText{
    _progressLabelText = progressLabelText;
    self.progressLabel.text = progressLabelText;
}

-(void)setProgressLabelTextColor:(UIColor *)progressLabelTextColor{
    _progressLabelTextColor = progressLabelTextColor;
    self.progressLabel.textColor = progressLabelTextColor;
}


#pragma mark - Draw CircleLine
- (void)setBackgroundCircleLine {
    UIBezierPath *path = [UIBezierPath bezierPath];
    path = [UIBezierPath bezierPathWithArcCenter:CGPointMake(self.center.x - self.frame.origin.x, self.center.y - self.frame.origin.y)
                                          radius:(self.frame.size.width - _backgroundLineWidth)/2 - _offset
                                      startAngle:self.startAngle
                                        endAngle:self.endAngle
                                       clockwise:NO];
    _backgroundLayer.path = path.CGPath;
}

- (void)setProgressCircleLine {
    UIBezierPath *path = [UIBezierPath bezierPath];
    path = [UIBezierPath bezierPathWithArcCenter:CGPointMake(self.center.x - self.frame.origin.x, self.center.y - self.frame.origin.y)
                                          radius:(self.frame.size.width - _progressLineWidth)/2 - _offset
                                      startAngle:self.startAngle
                                        endAngle:self.endAngle
                                       clockwise:NO];
    _progressLayer.path = path.CGPath;
}

#pragma mark - Lazy Load
- (UILabel *)progressLabel {
    if (!_progressLabel) {
        _progressLabel = [[UILabel alloc]initWithFrame:CGRectMake((self.bounds.size.width -100)/2, (self.bounds.size.height - 100)/2, 100, 100)];
    }
    return _progressLabel;
}

- (void)setBackgroundLineWidth:(CGFloat)backgroundLineWidth {
    _backgroundLineWidth = backgroundLineWidth;
    _backgroundLayer.lineWidth = _backgroundLineWidth;
    [self setBackgroundCircleLine];
}

-(void)setLineCap:(CAShapeLayerLineCap)lineCap{
    _progressLayer.lineCap = lineCap;
    [self setProgressCircleLine];
}

- (void)setProgressLineWidth:(CGFloat)progressLineWidth
{
    _progressLineWidth = progressLineWidth;
    _progressLayer.lineWidth = _progressLineWidth;
    [self setProgressCircleLine];
}

- (void)setPercentage:(CGFloat)percentage {
    _percentage = percentage;
}

- (void)setBackgroundStrokeColor:(UIColor *)backgroundStrokeColor {
    _backgroundStrokeColor = backgroundStrokeColor;
    _backgroundLayer.strokeColor = _backgroundStrokeColor.CGColor;
}

- (void)setProgressStrokeColor:(UIColor *)progressStrokeColor {
    _progressStrokeColor = progressStrokeColor;
    _progressLayer.strokeColor = _progressStrokeColor.CGColor;
}

#pragma mark - progress animated YES or NO
- (void)setProgress:(CGFloat)percentage animated:(BOOL)animated {
    self.percentage = percentage;
    _progressLayer.strokeEnd = _percentage;
    if (animated) {
        CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
        animation.fromValue = [NSNumber numberWithFloat:0.0];
        animation.toValue = [NSNumber numberWithFloat:_percentage];
        animation.duration = kDuration;
        [_progressLayer addAnimation:animation forKey:@"strokeEndAnimation"];
    }else{
        [CATransaction begin];
        [CATransaction setDisableActions:YES];
        _progressLabel.text = [NSString stringWithFormat:@"%.0f%%",_percentage*100];
        [CATransaction commit];
    }
}

@end

File: CollectionViewTest/CollectionViewTest/HuanView.swift
//
//  Huan.swift
//  CollectionViewTest
//
//  Created by STL_QYH on 2023/7/27.
//

import UIKit

class HuanView: UIView {

    override init(frame: CGRect) {
        super.init(frame: frame)
//        backgroundColor = .black
        let bPath: UIBezierPath = .init(arcCenter: CGPoint(x: center.x - frame.minX, y: center.y - frame.minY),
                                        radius: bounds.width / 2 - 5,
                                        startAngle: -Double.pi/2,
                                        endAngle: -Double.pi/2-2*Double.pi,
                                        clockwise: false)
        let bgLayer: CAShapeLayer = .init()
        bgLayer.strokeColor = UIColor.lightGray.cgColor
        bgLayer.path = bPath.cgPath
        bgLayer.fillColor = nil
        bgLayer.lineWidth = 10
//        bgLayer.frame = bounds
        layer.addSublayer(bgLayer)
        
        let lPath: UIBezierPath = .init(arcCenter: frame.origin,
                                        radius: bounds.width / 2 - 5,
                                        startAngle: -Double.pi/2,
                                        endAngle: -Double.pi/2-Double.pi,
                                        clockwise: false)
        let lLayer: CAShapeLayer = .init()
        lLayer.strokeColor = UIColor.orange.cgColor
        lLayer.path = lPath.cgPath
        lLayer.lineCap = .round
        lLayer.fillColor = nil
        lLayer.lineWidth = 10
//        lLayer.frame = bounds
        layer.addSublayer(lLayer)
        
        lLayer.strokeEnd = 0.95
        
        var point: CGPoint = bounds.origin
        point.x += bounds.width / 2.0
        point.y += bounds.height / 2.0
        let rPath: UIBezierPath = .init(arcCenter: point,
                                        radius: bounds.width / 2 - 5,
                                        startAngle: -Double.pi/2,
                                        endAngle: -Double.pi/2-Double.pi,
                                        clockwise: true)
        let rLayer: CAShapeLayer = .init()
        rLayer.strokeColor = UIColor.orange.cgColor
        rLayer.path = rPath.cgPath
        rLayer.lineCap = .round
        rLayer.fillColor = nil
        rLayer.lineWidth = 10
//        rLayer.frame = bounds
        layer.addSublayer(rLayer)
        
        rLayer.strokeEnd = 0.95
        
        let animation: CABasicAnimation = .init(keyPath: "strokeEnd")
        animation.fromValue = 0
        animation.toValue = 0.95
        animation.duration = 1
        
        lLayer.add(animation, forKey: "strokeEndAnimation")
        rLayer.add(animation, forKey: "strokeEndAnimation")
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */
}
